core/utils/toolbox.js

```diff
@@ -135,75 +136,126 @@ Blockly.utils.toolbox.CATEGORY_TOOLBOX_KIND = 'categoryToolbox';
+
+  if (toolboxDef instanceof Element || typeof toolboxDef == 'string') {
+    toolboxDef = Blockly.utils.toolbox.parseToolboxTree(toolboxDef);
+    toolboxDef = Blockly.utils.toolbox.convertToToolboxJson_(toolboxDef);
   }
+
+  var toolboxJson = /** @type {Blockly.utils.toolbox.ToolboxJson} */ (toolboxDef);
+  Blockly.utils.toolbox.validateToolbox_(toolboxJson);
   return toolboxJson;
 };
 
 /**
- * Converts the toolbox contents to JSON.
- * @param {Blockly.utils.toolbox.ToolboxDefinition} toolboxDef The definition
- *     of the toolbox in one of its many forms.
- * @return {Array<Blockly.utils.toolbox.FlyoutItem>|
- *          Array<Blockly.utils.toolbox.ToolboxItem>} The contents of the
- *          toolbox.
+ * Validates the toolbox JSON fields have been set correctly.
+ * @param {Blockly.utils.toolbox.ToolboxJson} toolboxJson Object holding
+ *     information for creating a toolbox.
+ * @throws {Error} if the toolbox is not the correct format.
+ * @private
+ */
+Blockly.utils.toolbox.validateToolbox_ = function(toolboxJson){
+  var toolboxKind = toolboxJson['kind'];
+  var toolboxContents = toolboxJson['contents'];
+
+  if (toolboxKind) {
+    if (toolboxKind != Blockly.utils.toolbox.FLYOUT_TOOLBOX_KIND &&
+      toolboxKind != Blockly.utils.toolbox.CATEGORY_TOOLBOX_KIND) {
+      throw Error('Invalid toolbox kind ' + toolboxKind + '.' +
+        ' Please supply either ' +
+        Blockly.utils.toolbox.FLYOUT_TOOLBOX_KIND + ' or ' +
+        Blockly.utils.toolbox.CATEGORY_TOOLBOX_KIND);
+    }
+  }
+  if (!toolboxContents) {
+    throw Error('Toolbox must have a contents attribute.');
+  }
+};
+
+/**
+ * Converts the flyout definition into a list of flyout items.
+ * @param {?Blockly.utils.toolbox.FlyoutDefinition} flyoutDef The definition of
+ *    the flyout in one of its many forms.
+ * @return {!Array<Blockly.utils.toolbox.FlyoutItemJson>} A list of flyout items.
  * @package
  */
-Blockly.utils.toolbox.convertToolboxContentsToJSON = function(toolboxDef) {
-  if (!toolboxDef) {
-    return null;
+Blockly.utils.toolbox.convertFlyoutDefToJsonArray = function(flyoutDef) {
+  if (!flyoutDef) {
+    return [];
   }
-  var contents = toolboxDef['contents'] || toolboxDef;
-  // If it is an array of JSON, then it is already in the correct format.
-  if (Blockly.utils.toolbox.isCorrectFormat(contents)) {
-    if (Blockly.utils.toolbox.hasCategories(toolboxDef)) {
-      // TODO: Remove after #3985 has been looked into.
-      console.warn('Due to some performance issues, defining a toolbox using' +
-          ' JSON is not ready yet. Please define your toolbox using xml.');
-    }
-    return contents;
+
+  if (flyoutDef['contents']) {
+    return flyoutDef['contents'];
+  }
+
+  // If it is already in the correct format return the flyoutDef.
+  if (Array.isArray(flyoutDef) && flyoutDef.length > 0 &&
+      !flyoutDef[0].nodeType) {
+    return flyoutDef;
   }
 
   return Blockly.utils.toolbox.toolboxXmlToJson_(
```

Comment:
Why is this still `toolboxXmlToJson_` if this is guaranteed to be a flyout?

Reply:
Updated to xmlToJsonArray_